service/share/share.go

```diff
@@ -161,37 +161,37 @@ func (s *service) GetSharesByNamespace(ctx context.Context, root *Root, nID name
 		return nil, ipld.ErrNotFoundInRange
 	}
 
-	type res struct {
-		nodes []format.Node
-		err   error
-	}
-	resultCh := make(chan *res)
-
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	for _, rootCID := range rowRootCIDs {
-		go func(rootCID cid.Cid) {
-			nodes, err := ipld.GetLeavesByNamespace(ctx, s.dag, rootCID, nID)
-			resultCh <- &res{nodes: nodes, err: err}
-		}(rootCID)
+	errCh := make(chan error)
+	nodes := make([][]format.Node, len(rowRootCIDs))
+	for i, rootCID := range rowRootCIDs {
+		go func(i int, rootCID cid.Cid) {
+			var err error
+			nodes[i], err = ipld.GetLeavesByNamespace(ctx, s.dag, rootCID, nID)
+			errCh <- err
+		}(i, rootCID)
 	}
 
-	namespacedShares := make([]Share, 0)
 	for i := 0; i < len(rowRootCIDs); i++ {
 		select {
-		case result := <-resultCh:
-			if result.err != nil {
-				return nil, result.err
-			}
-			for _, node := range result.nodes {
-				namespacedShares = append(namespacedShares, node.RawData()[1:])
+		case err := <-errCh:
+			if err != nil {
+				return nil, err
 			}
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		}
 	}
 
+	namespacedShares := make([]Share, 0)
```

Comment:
Preallocate capacity pls

Reply:
Each node can have arbitrary number of shares inside. So without iterating, can't per-allocate _precisely_. Another option is to push number of shares along with the error using a channel. Both adds complexity.
I can also pre-allocate for `len(nodes)` aka `len(rowRootCIDs)`.

Reply:
Oh, true, we never know precisely how many shares we submit unless we encode the number of shares in the first metadata like share when PFD is sent. 

It is fine for now to keep it as it is then.